src/proxyp/Parser.cc

```diff
@@ -0,0 +1,287 @@
+        throw TexcHere("PROXY/1.0 error: missing SP after the magic sequence");
+
+    static const SBuf protoUnknown("UNKNOWN");
+    static const SBuf protoTcp("TCP");
+
+    if (interiorTok.skip(protoTcp))
+        parseAddresses(interiorTok, message);
+    else if (interiorTok.skip(protoUnknown))
+        message->ignoreAddresses();
+    else
+        throw TexcHere("PROXY/1.0 error: invalid INET protocol or family");
+
+    return Parsed(message, tok.parsedSize());
+}
+
+void
+ProxyProtocol::Two::parseAddresses(const uint8_t family, Parser::BinaryTokenizer &tok, MessagePointer &message)
+{
+    switch (family) {
+
+    case afInet: {
+        message->sourceAddress = tok.inet4("src_addr IPv4");
+        message->destinationAddress = tok.inet4("dst_addr IPv4");
+        message->sourceAddress.port(tok.uint16("src_port"));
+        message->destinationAddress.port(tok.uint16("dst_port"));
+        break;
+    }
+
+    case afInet6: {
+        message->sourceAddress = tok.inet6("src_addr IPv6");
+        message->destinationAddress = tok.inet6("dst_addr IPv6");
+        message->sourceAddress.port(tok.uint16("src_port"));
+        message->destinationAddress.port(tok.uint16("dst_port"));
+        break;
+    }
+
+    case afUnix: { // TODO: add support
+        // the address block length is 216 bytes
+        tok.skip(216, "unix_addr");
+        break;
+    }
+
+    default: {
+        // unreachable code: we have checked family validity already
+        Must(false);
+        break;
+    }
+    }
+}
+
+void
+ProxyProtocol::Two::parseTLVs(Parser::BinaryTokenizer &tok, MessagePointer &message) {
+    while (!tok.atEnd()) {
+        const auto type = tok.uint8("pp2_tlv::type");
+        message->tlvs.emplace_back(type, tok.pstring16("pp2_tlv length and value"));
+    }
+}
+
+ProxyProtocol::Parsed
+ProxyProtocol::Two::Parse(const SBuf &buf)
+{
+    Parser::BinaryTokenizer tokMessage(buf, true);
+
+    const auto versionAndCommand = tokMessage.uint8("version and command");
+
+    const auto version = (versionAndCommand & 0xF0) >> 4;
+    if (version != 2) // version == 2 is mandatory
+        throw TexcHere(ToSBuf("PROXY/2.0 error: invalid version ", version));
+
+    const auto command = (versionAndCommand & 0x0F);
+    if (command > cmdProxy)
+        throw TexcHere(ToSBuf("PROXY/2.0 error: invalid command ", command));
+
+    const auto familyAndProto = tokMessage.uint8("family and proto");
+
+    const auto family = (familyAndProto & 0xF0) >> 4;
+    if (family > afUnix)
+        throw TexcHere(ToSBuf("PROXY/2.0 error: invalid address family ", family));
+
+    const auto proto = (familyAndProto & 0x0F);
+    if (proto > tpDgram)
+        throw TexcHere(ToSBuf("PROXY/2.0 error: invalid transport protocol ", proto));
+
+    // the header length field contains the number following bytes beyond this field
+    const auto headerLen = tokMessage.uint16("header length");
+
+    const auto header = tokMessage.area(headerLen, "header");
```

Comment:
```suggestion
    const auto header = tokMessage.pstring16("header");
```

And drop headerLen above, along with its comment.